File: api-exam/conference-service/src/main/java/ma/enset/conferenceservice/services/ConferenceService.java
package ma.enset.conferenceservice.services;

import ma.enset.conferenceservice.dtos.ConferenceDTO;
import ma.enset.conferenceservice.enums.ConferenceType;

import java.util.List;

public interface ConferenceService {
    ConferenceDTO saveConference(ConferenceDTO conferenceDTO);

    ConferenceDTO updateConference(Long id, ConferenceDTO conferenceDTO);

    ConferenceDTO getConferenceById(Long id);

    ConferenceDTO getConferenceByIdWithKeynote(Long id);

    List<ConferenceDTO> getAllConferences();

    List<ConferenceDTO> getAllConferencesWithKeynotes();

    void deleteConference(Long id);

    List<ConferenceDTO> getConferencesByType(ConferenceType type);

    List<ConferenceDTO> searchByTitre(String titre);

    List<ConferenceDTO> getConferencesByKeynoteId(Long keynoteId);
}

File: ui-exam/src/app/services/conference.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Conference, ConferenceType, Review } from '../models';
import { environment } from '../../environments/environment';

@Injectable({
    providedIn: 'root'
})
export class ConferenceService {
    private apiUrl = `${environment.apiUrl}/api/conferences`;

    constructor(private http: HttpClient) { }

    getAllConferences(withKeynotes: boolean = true): Observable<Conference[]> {
        return this.http.get<Conference[]>(`${this.apiUrl}?withKeynotes=${withKeynotes}`);
    }

    getConferenceById(id: number, withKeynote: boolean = true): Observable<Conference> {
        return this.http.get<Conference>(`${this.apiUrl}/${id}?withKeynote=${withKeynote}`);
    }

    createConference(conference: Conference): Observable<Conference> {
        return this.http.post<Conference>(this.apiUrl, conference);
    }

    updateConference(id: number, conference: Conference): Observable<Conference> {
        return this.http.put<Conference>(`${this.apiUrl}/${id}`, conference);
    }

    deleteConference(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    getConferencesByType(type: ConferenceType): Observable<Conference[]> {
        return this.http.get<Conference[]>(`${this.apiUrl}/type/${type}`);
    }

    searchByTitre(titre: string): Observable<Conference[]> {
        return this.http.get<Conference[]>(`${this.apiUrl}/search?titre=${titre}`);
    }

    getConferencesByKeynoteId(keynoteId: number): Observable<Conference[]> {
        return this.http.get<Conference[]>(`${this.apiUrl}/keynote/${keynoteId}`);
    }

    // Review methods
    getReviewsByConferenceId(conferenceId: number): Observable<Review[]> {
        return this.http.get<Review[]>(`${this.apiUrl}/${conferenceId}/reviews`);
    }

    addReview(conferenceId: number, review: Review): Observable<Review> {
        return this.http.post<Review>(`${this.apiUrl}/${conferenceId}/reviews`, review);
    }
}
